File: wdOS.SDK/librt/userconsole.cpp
#include "../include/wdos.hpp"
using namespace wdOS;

ushort UserConsole::_tableid = 0xFF04;

uint UserConsole::GetCursorX() {
	return PlatformSyscall(_tableid, 0, 0);
}
uint UserConsole::GetCursorY() {
	return PlatformSyscall(_tableid, 1, 0);
}
uint UserConsole::GetConsoleForeground() {
	return PlatformSyscall(_tableid, 2, 0);
}
uint UserConsole::GetConsoleBackground() {
	return PlatformSyscall(_tableid, 3, 0);
}
void UserConsole::SetConsoleForeground(uint val) {
	PlatformSyscall(_tableid, 4, val);
}
void UserConsole::SetConsoleBackground(uint val) {
	PlatformSyscall(_tableid, 5, val);
}
void UserConsole::SetCursorX(uint val) {
	PlatformSyscall(_tableid, 6, val);
}
void UserConsole::SetCursorY(uint val) {
	PlatformSyscall(_tableid, 7, val);
}
void UserConsole::Write(char* buffer) {
	PlatformSyscall(_tableid, 8, (uint)buffer);
}
uint UserConsole::Read(char* buffer) {
	return PlatformSyscall(_tableid, 9, (uint)buffer);
}

File: wdOS.Platform/UserConsole.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace wdOS.Platform
{
    // only for user programs
    internal unsafe static class UserConsole
    {
        internal static uint GetCursorX() => (uint)Console.CursorLeft;
        internal static uint GetCursorY() => (uint)Console.CursorTop;
        internal static uint GetConsoleForeground() => (uint)Console.ForegroundColor;
        internal static uint GetConsoleBackground() => (uint)Console.BackgroundColor;
        internal static void SetConsoleForeground(uint val) => Console.ForegroundColor = (ConsoleColor)val;
        internal static void SetConsoleBackground(uint val) => Console.BackgroundColor = (ConsoleColor)val;
        internal static void SetCursorX(uint val) => Console.SetCursorPosition((int)val, Console.CursorTop);
        internal static void SetCursorY(uint val) => Console.SetCursorPosition(Console.CursorLeft, (int)val);
        internal static void Write(char* buffer)
        {
            for (int i = 0; buffer[i] != 0; i++)
                Console.WriteLine(buffer[i]);
        }
        internal static uint Read(char* buffer)
        {
            string str = Console.ReadLine();
            for (int i = 0; i < str.Length; i++)
                buffer[i] = str[i];
            buffer[str.Length] = '\0';
            return (uint)str.Length;
        }
    }
}
